Format string error in WaveShare/Goodisplay display driver"

### The problem

The fix would be simple : changing line 3135 in waveshare_epaper.cpp from %d to %lu.

src/esphome/components/waveshare_epaper/waveshare_epaper.cpp:3135:19: error: format '%d' expects argument of type 'int', but argument 5 has type 'long unsigned int' [-Werror=format=]
 3135 |     ESP_LOGD(TAG, "Partial update done, next full update after %d cycles",
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/core/log.h:76:36: note: in definition of macro 'ESPHOME_LOG_FORMAT'
   76 | #define ESPHOME_LOG_FORMAT(format) format
      |                                    ^~~~~~
src/esphome/core/log.h:156:28: note: in expansion of macro 'esph_log_d'
  156 | #define ESP_LOGD(tag, ...) esph_log_d(tag, __VA_ARGS__)
      |                            ^~~~~~~~~~
src/esphome/components/waveshare_epaper/waveshare_epaper.cpp:3135:5: note: in expansion of macro 'ESP_LOGD'
 3135 |     ESP_LOGD(TAG, "Partial update done, next full update after %d cycles",
      |     ^~~~~~~~
src/esphome/components/waveshare_epaper/waveshare_epaper.cpp:3135:65: note: format string is defined here
 3135 |     ESP_LOGD(TAG, "Partial update done, next full update after %d cycles",
      |                                                                ~^
      |                                                                 |
      |                                                                 int
      |                                                                %ld

### Which version of ESPHome has the issue?

 ESPHome 2025.6.1

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

latest

### What platform are you using?

ESP32

### Board

esp32-c6-devkitc-1

### Component causing the issue

waveshare_epaper/waveshare_epaper.cpp

### YAML Config

```yaml
esphome:
  name: kpis
  friendly_name: kpis
  on_boot:
    priority: -10  # Lower priority to allow Wi-Fi and API to connect first
    then:
      - wait_until:
          condition:
            and:
              - wifi.connected
              - api.connected
      - script.execute: update_screen

esp32:
  board: esp32-c6-devkitc-1
  flash_size: 4MB
  variant: esp32c6
  framework:
    type: esp-idf
    version: "5.3.1"
    platform_version: 6.9.0
    sdkconfig_options:
      CONFIG_ESPTOOLPY_FLASHSIZE_4MB: y


# Enable logging
logger:
  level: DEBUG

# Enable Home Assistant API
api:
  encryption:
    key: "og/sh4/D9Uc/zUVAp2xSsgitQ8loscOW/KMGApRttrI="

ota:
  - platform: esphome
    password: "1984d4461004e5ea160863e2bbdc2ce8"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  fast_connect: true
#  output_power: 16dB # Uncomment if you have unstable WiFi, this reduces power drew by the WiFi chip

  # Enable fallback hotspot (captive portal) in case wifi connection fails
#  ap:
#    ssid: "Fallback Hotspot"
#    password: "whx8K34CJ23L"

captive_portal:

external_components:
  - source:
      # https://github.com/lboue/esphome/blob/adc_oneshot/esphome/components/adc
      type: git
      url: https://github.com/lboue/esphome
      ref: adc_oneshot
    components: [ adc ]
    refresh: 0s
  
# Built-in button on GPIO9
binary_sensor:
#  - platform: gpio
#    pin:
#      number: GPIO9
#      mode: INPUT_PULLUP
#      inverted: true
#    name: "C6 Button"
#    filters:
#      - delayed_on: 10ms
#      - delayed_off: 10ms

  # Will only pick up a value if this device is configured in Home Assistant > Integrations
  # If the device isn't configured, or Home Assistant is offline, it'll default to false
  - platform: homeassistant
    id: dashboard_prevent_deep_sleep
    entity_id: input_boolean.dashboard_prevent_deep_sleep



script:
  - id: update_screen
    mode: queued
    then:
      - wait_until:
          condition:
            and:
              - lambda: 'return !id(p1_meter_power).state.empty();'
              - lambda: 'return !id(energy_socket_power_3).state.empty();'
              - lambda: 'return !id(boost_state).state.empty();'
              - lambda: 'return !id(temp_reading).state.empty();'
              - lambda: 'return !id(bambulab).state.empty();'
              - lambda: 'return !id(warmingwall).state.empty();'
              - lambda: 'return !id(next_car_power).state.empty();'
              - lambda: 'return !id(next_car_power_mode).state.empty();'
      - component.update: eink_display
      - if:
          condition:
            binary_sensor.is_on: dashboard_prevent_deep_sleep
          then:
            - logger.log: '*-*-*-*-*-*-*-*-*-SKIPPING DEEP SLEEP*-*-*-*-*-*-*-*-*-*-*-*'
           
            - logger.log: 
                format: "*-*-*-*-*-*-*-*-*-PAUSE DELAY %s *-*-*-*-*-*-*-*-*-"
                args: [ 'id(dashboard_delay_refresh).state.c_str()' ]
            
            - delay: !lambda |-
                // Convert the string to a float
                float delay_seconds = atof(id(dashboard_delay_refresh).state.c_str())*60*1000;
                return delay_seconds;

          else:
            - logger.log: '*-*-*-*-*-*-*-*-*-DEEP SLEEP*-*-*-*-*-*-*-*-*-'
            - deep_sleep.enter: deep_sleep_id
      - script.execute: update_screen

time:
  - platform: sntp
    id: sntp_time

spi:
  clk_pin: 23 #CLK #SCK
  mosi_pin: 22 #DIN SDI
    
font:
  ######### REGULAR #################
  - file: "fonts/GothamRnd-Book.ttf"
    id: book20
    size: 20
    glyphs: &font-glyphs
      ['!', '"', '%', '(', ')', '+', '=', ',', '-', '_', '.', ':', '°', ' ',
      '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
      'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
      'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
      'é', 'à', 'è', 'ç', 'ù', 'É', 'À', 'È', 'Ç', 'Ù']
  - file: "fonts/GothamRnd-Book.ttf"
    id: book25
    size: 25
    glyphs : *font-glyphs 
  ########## BOLD ################## 
  - file: "fonts/GothamRnd-Bold.ttf"
    id: bold15
    size: 15
    glyphs : *font-glyphs 
  - file: "fonts/GothamRnd-Bold.ttf"
    id: bold20
    size: 20
    glyphs : *font-glyphs 
  - file: "fonts/GothamRnd-Bold.ttf"
    id: bold25
    size: 25
    glyphs : *font-glyphs 
  - file: "fonts/GothamRnd-Bold.ttf"
    id: bold30
    size: 30
    glyphs : *font-glyphs 
  - file: "fonts/GothamRnd-Bold.ttf"
    id: bold35
    size: 35
    glyphs : *font-glyphs 
  - file: "fonts/GothamRnd-Bold.ttf"
    id: bold40
    size: 40
    glyphs : *font-glyphs 
  - file: "fonts/GothamRnd-Bold.ttf"
    id: bold55
    size: 55
    glyphs: ['.', '°', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'C', '-']
  
  - file: 'fonts/materialdesignicons-webfont.ttf'
    id: mdi25
    size: 25
    glyphs:
        - "\U000F0425"  # mdi-flash
        - "\U000F0F92"  # mdi-heater (alternative to mdi-water-boiler)
        - "\U000F042B"  # mdi-printer-3d
        - "\U000F0438"  # mdi-wall
        - "\U000F0B6C"  # mdi-car-electric
        - "\U000F007A"  # mdi-battery-10
        - "\U000F007B"  # mdi-battery-20
        - "\U000F007C"  # mdi-battery-30
        - "\U000F007D"  # mdi-battery-40
        - "\U000F007E"  # mdi-battery-50
        - "\U000F007F"  # mdi-battery-60
        - "\U000F0080"  # mdi-battery-70
        - "\U000F0081"  # mdi-battery-80
        - "\U000F0082"  # mdi-battery-90
        - "\U000F0079"  # mdi-battery-100

sensor:
  - platform: wifi_signal
    name: "WiFi Signal"
    update_interval: 5s
  - platform: uptime
    name: "Uptime"
  - name: "Battery voltage"
    platform: adc
    id: batt_volt
    pin: GPIO0
    update_interval: 5s
    filters:
      - multiply: 2
  - platform: template
    name: "Battery level"
    state_class: "measurement"
    device_class: "battery"
    unit_of_measurement: "%"
    update_interval: 5s
    accuracy_decimals: 0
    lambda: |-
      return round(((id(batt_volt).state-3) /1.130 * 100.00));


text_sensor: 
  - platform: homeassistant
    entity_id: sensor.energy_socket_power_3
    id: energy_socket_power_3
  - platform: homeassistant
    entity_id: sensor.p1_meter_power
    id: p1_meter_power

  - platform: homeassistant
    entity_id: sensor.next_car_power_mode
    id: next_car_power_mode

  - platform: homeassistant
    entity_id: sensor.next_car_power
    id: next_car_power

  - platform: homeassistant
    entity_id: sensor.energy_socket_power_2
    id: bambulab

  - platform: homeassistant
    entity_id: sensor.energy_socket_power
    id: warmingwall

  - platform: homeassistant
    id: dashboard_delay_refresh
    entity_id: input_text.dashboard_delay_refresh

  - platform: homeassistant
    entity_id: sensor.temp_reading
    id: temp_reading
  - platform: homeassistant
    entity_id: sensor.boost_state
    id: boost_state 

#graph:
#  - id: weather_temperature_graph
#    sensor: weather_temperature
#    duration: 24h
#    width: 380
#    height: 50
#    min_value: 0

deep_sleep:
  id: deep_sleep_id
  #run_duration: 30s
  sleep_duration: 10min
  wakeup_pin:
    number: GPIO4
    mode: INPUT_PULLUP
    inverted: true


display:
  - platform: waveshare_epaper
    id: eink_display
    cs_pin: 17
    dc_pin: 16
    busy_pin: 14
    reset_pin: 21
    reset_duration: 2ms
    model: 7.50inv2p  # Changed from 7.50inv2
    rotation: 0°
    update_interval: never
    full_update_every: 30  # Full refresh every 30 partial updates


  

    lambda: |
     // Display header
      it.printf(30, 10, id(bold35), TextAlign::LEFT, "%s", "Dashboard");
      it.filled_rectangle(30, 45, (800 - 60), 2);

      // Define positions
      int x_icon = 30;
      int x_text = x_icon + 35;  // Leave space for icon
      int y_start = 60;
      int line_height = 45;      // Adjust as needed

      // Line 1: Total Power Consumption
      it.print(x_icon, y_start, id(mdi25), "\U000F0425");  // mdi-flash
      it.printf(x_text, y_start, id(bold30), TextAlign::LEFT, "Total Power: %d W", (int) round(atof(id(p1_meter_power).state.c_str())));

      // Line 2: Boiler Power
      it.print(x_icon, y_start + line_height, id(mdi25), "\U000F0F92");  // mdi-water-boiler
      it.printf(x_text, y_start + line_height, id(bold30), TextAlign::LEFT, "Boiler Power: %d W (Boost: %s)", (int) round(atof(id(energy_socket_power_3).state.c_str())), id(boost_state).state.c_str() );

      // Line 3: Boiler Temp
      it.print(x_icon, y_start + 2 * line_height, id(mdi25), "\U000F0F92");  // mdi-printer-3d
      it.printf(x_text, y_start + 2 * line_height, id(bold30), TextAlign::LEFT, "Boiler Temperature: %.1f C",  atof(id(temp_reading).state.c_str()));

      // Line 4: Bambulab Power
      it.print(x_icon, y_start + 3 * line_height, id(mdi25), "\U000F042B");  // mdi-printer-3d
      it.printf(x_text, y_start + 3 * line_height, id(bold30), TextAlign::LEFT, "Bambulab Power: %d W", (int) round(atof(id(bambulab).state.c_str())));

      // Line 4: Wall Power
      it.print(x_icon, y_start + 4 * line_height, id(mdi25), "\U000F0438");  // mdi-wall
      it.printf(x_text, y_start + 4 * line_height, id(bold30), TextAlign::LEFT, "Wall Power: %d W", (int) round(atof(id(warmingwall).state.c_str())));

      // Line 5: Car Consumption
      it.print(x_icon, y_start + 5 * line_height, id(mdi25), "\U000F0B6C");  // mdi-car-electric
      it.printf(x_text, y_start + 5 * line_height, id(bold30), TextAlign::LEFT, "Car Consumption: %d W (%s)", (int) atof(id(next_car_power).state.c_str()), id(next_car_power_mode).state.c_str());

      // Calculate battery percentage
      float battery_percent = (id(batt_volt).state - 3) / 1.125 * 100;
      const char* battery_icon;

      if (battery_percent >= 90)
        battery_icon = "\U000F0079";  // mdi-battery-100
      else if (battery_percent >= 80)
        battery_icon = "\U000F0082";  // mdi-battery-90
      else if (battery_percent >= 70)
        battery_icon = "\U000F0081";  // mdi-battery-80
      else if (battery_percent >= 60)
        battery_icon = "\U000F0080";  // mdi-battery-70
      else if (battery_percent >= 50)
        battery_icon = "\U000F007F";  // mdi-battery-60
      else if (battery_percent >= 40)
        battery_icon = "\U000F007E";  // mdi-battery-50
      else if (battery_percent >= 30)
        battery_icon = "\U000F007D";  // mdi-battery-40
      else if (battery_percent >= 20)
        battery_icon = "\U000F007C";  // mdi-battery-30
      else if (battery_percent >= 10)
        battery_icon = "\U000F007B";  // mdi-battery-20
      else
        battery_icon = "\U000F007A";  // mdi-battery-10

      // Display battery icon and percentage at the bottom left
      it.print(10, 480, id(mdi25), TextAlign::BOTTOM_LEFT, battery_icon);
      it.printf(45, 480, id(bold15), TextAlign::BOTTOM_LEFT, "Battery: %.1f%%", battery_percent);

      // Display update info at bottom right
      it.printf(660, 480, id(bold15), TextAlign::BOTTOM_RIGHT, "Updated: ");
      it.strftime(800, 480, id(bold15), TextAlign::BOTTOM_RIGHT, "%d-%m-%Y %H:%M", id(sntp_time).now());
```

### Anything in the logs that might be useful for us?

```txt

```

### Additional information

_No response_

--- a/esphome/components/waveshare_epaper/waveshare_epaper.cpp
+++ b/esphome/components/waveshare_epaper/waveshare_epaper.cpp
@@ -3132,7 +3132,7 @@ void HOT GDEY0583T81::display() {
   } else {
     // Partial out (PTOUT), makes the display exit partial mode
     this->command(0x92);
-    ESP_LOGD(TAG, "Partial update done, next full update after %lu cycles",
+    ESP_LOGD(TAG, "Partial update done, next full update after %" PRIu32 " cycles",
              this->full_update_every_ - this->at_update_ - 1);
   }
 

--- a/esphome/components/waveshare_epaper/waveshare_epaper.cpp
+++ b/esphome/components/waveshare_epaper/waveshare_epaper.cpp
@@ -3132,7 +3132,7 @@ void HOT GDEY0583T81::display() {
   } else {
     // Partial out (PTOUT), makes the display exit partial mode
     this->command(0x92);
-    ESP_LOGD(TAG, "Partial update done, next full update after %d cycles",
+    ESP_LOGD(TAG, "Partial update done, next full update after %lu cycles",
              this->full_update_every_ - this->at_update_ - 1);
   }
 

